File: tests/integration/fixtures/relay.py
import json
import os
import sys
import signal
import stat
import requests
import subprocess

import pytest

from . import SentryLike

RELAY_BIN = [os.environ.get("RELAY_BIN") or "target/debug/relay"]

if os.environ.get("RELAY_AS_CARGO", "false") == "true":
    RELAY_BIN = ["cargo", "run", "--"]


class Relay(SentryLike):
    def __init__(
        self,
        server_address,
        process,
        upstream,
        public_key,
        secret_key,
        relay_id,
        config_dir,
        options,
        version,
    ):
        super(Relay, self).__init__(server_address, upstream, public_key)

        self.process = process
        self.relay_id = relay_id
        self.secret_key = secret_key
        self.config_dir = config_dir
        self.options = options
        self.version = version

    def shutdown(self, sig=signal.SIGKILL):
        self.process.send_signal(sig)

        try:
            self.process.wait(19)
        except subprocess.TimeoutExpired:
            self.process.kill()
            raise


@pytest.fixture
def get_relay_binary():
    def inner(version="latest"):
        if version == "latest":
            return RELAY_BIN

        if sys.platform == "linux" or sys.platform == "linux2":
            filename = "relay-Linux-x86_64"
        elif sys.platform == "darwin":
            filename = "relay-Darwin-x86_64"
        elif sys.platform == "win32":
            filename = "relay-Windows-x86_64.exe"

        download_path = f"target/relay_releases_cache/{filename}_{version}"

        if not os.path.exists(download_path):
            download_url = (
                f"https://github.com/getsentry/relay/releases/download/"
                f"{version}/{filename}"
            )

            headers = {}
            if "GITHUB_TOKEN" in os.environ:
                headers["Authorization"] = f"Bearer {os.environ['GITHUB_TOKEN']}"

            os.makedirs(os.path.dirname(download_path), exist_ok=True)

            with requests.get(download_url, headers=headers) as r:
                r.raise_for_status()

                with open(download_path, "wb") as f:
                    for chunk in r.iter_content(chunk_size=8192):
                        f.write(chunk)

        os.chmod(download_path, 0o700 | stat.S_IEXEC)

        return [download_path]

    return inner


@pytest.fixture
def relay(mini_sentry, random_port, background_process, config_dir, get_relay_binary):
    def inner(
        upstream,
        options=None,
        prepare=None,
        external=None,
        wait_healthcheck=True,
        static_relays=None,
        version="latest",
    ):
        relay_bin = get_relay_binary(version)
        host = "127.0.0.1"
        port = random_port()

        default_opts = {
            "relay": {
                "upstream": upstream.url,
                "host": host,
                "port": port,
                "tls_port": None,
                "tls_private_key": None,
                "tls_cert": None,
            },
            "sentry": {"dsn": mini_sentry.internal_error_dsn, "enabled": True},
            "limits": {"max_api_file_upload_size": "1MiB"},
            "cache": {"batch_interval": 0},
            "logging": {"level": "trace"},
            "http": {"timeout": 2},
            "processing": {"enabled": False, "kafka_config": [], "redis": ""},
        }

        if static_relays is not None:
            default_opts["auth"] = {"static_relays": static_relays}

        if options is not None:
            for key in options:
                default_opts.setdefault(key, {}).update(options[key])

        dir = config_dir("relay")
        dir.join("config.yml").write(json.dumps(default_opts))

        output = subprocess.check_output(
            relay_bin + ["-c", str(dir), "credentials", "generate"]
        )

        # now that we have generated a credentials file get the details
        with open(dir.join("credentials.json"), "r") as f:
            credentials = json.load(f)
        public_key = credentials.get("public_key")
        assert public_key is not None
        secret_key = credentials.get("secret_key")
        assert secret_key is not None
        relay_id = credentials.get("id")
        assert relay_id is not None

        if prepare is not None:
            prepare(dir)

        mini_sentry.known_relays[relay_id] = {
            "publicKey": public_key,
            "internal": not external,
            "version": version,
        }

        process = background_process(relay_bin + ["-c", str(dir), "run"])

        relay = Relay(
            (host, port),
            process,
            upstream,
            public_key,
            secret_key,
            relay_id,
            dir,
            default_opts,
            version,
        )

        if wait_healthcheck:
            relay.wait_relay_healthcheck()

        return relay

    return inner

File: relay-test/src/relay.rs
use std::net::SocketAddr;
use std::path::{Path, PathBuf};
use std::sync::{Arc, PoisonError};

use hyper::http::HeaderName;
use std::time::Duration;

use axum::http::HeaderMap;
use relay_auth::{PublicKey, SecretKey};
use relay_base_schema::project::{ProjectId, ProjectKey};
use relay_config::Config;
use relay_config::Credentials;
use relay_config::RelayInfo;
use reqwest::{self, Response};
use serde_json::{json, Value};
use std::env;
use std::fs::{self, File};
use std::io::Write;
use std::os::unix::fs::PermissionsExt;
use tokio::runtime::Runtime;
use uuid::Uuid;

use crate::mini_sentry::MiniSentry;
use crate::{
    merge, outcomes_enabled_config, processing_config, random_port, BackgroundProcess, RawEnvelope,
    TempDir,
};

pub trait Upstream {
    fn url(&self) -> String;
    fn internal_error_dsn(&self) -> String;
    fn insert_known_relay(&self, relay_id: Uuid, public_key: PublicKey);
    fn public_dsn_key(&self, id: ProjectId) -> ProjectKey;
}

impl<U: Upstream> Upstream for Relay<'_, U> {
    fn url(&self) -> String {
        self.url()
    }

    fn internal_error_dsn(&self) -> String {
        self.upstream_dsn.clone()
    }

    fn insert_known_relay(&self, relay_id: Uuid, public_key: PublicKey) {
        self.upstream.insert_known_relay(relay_id, public_key);
    }

    fn public_dsn_key(&self, id: ProjectId) -> ProjectKey {
        self.upstream.public_dsn_key(id)
    }
}

impl Upstream for MiniSentry {
    fn url(&self) -> String {
        self.inner.lock().unwrap().url()
    }

    fn internal_error_dsn(&self) -> String {
        self.inner
            .lock()
            .unwrap_or_else(PoisonError::into_inner)
            .internal_error_dsn()
    }

    fn insert_known_relay(&self, relay_id: Uuid, public_key: PublicKey) {
        self.inner.lock().unwrap().known_relays.insert(
            relay_id,
            RelayInfo {
                public_key,
                internal: true,
            },
        );
    }

    fn public_dsn_key(&self, id: ProjectId) -> ProjectKey {
        self.get_dsn_public_key_configs(id).unwrap().public_key
    }
}

fn default_opts(url: String, internal_error_dsn: String, port: u16, host: String) -> Value {
    json!({
        "relay": {
            "upstream": url,
            "host": host,
            "port": port,
            "tls_port": null,
            "tls_private_key": null,
            "tls_cert": null,
        },
        "sentry": {
            "dsn": internal_error_dsn,
            "enabled": true,
        },
        "limits": {
            "max_api_file_upload_size": "1MiB",
        },
        "cache": {
            "batch_interval": 0,
        },
        "logging": {
            "level": "trace",
        },
        "http": {
            "timeout": 3,
        },
        "processing": {
            "enabled": false,
            "kafka_config": [],
            "redis": "",
        },
        "outcomes": {
            "aggregator": {
                "bucket_interval": 1,
                "flush_interval": 0,
            },
        },
    })
}

pub struct RelayBuilder<'a, U: Upstream> {
    pub config: serde_json::Value,
    upstream: &'a U,
}

impl<'a, U: Upstream> RelayBuilder<'a, U> {
    pub fn enable_processing(mut self) -> Self {
        let proc = processing_config();

        self.config = merge(self.config, proc, vec![]);
        self
    }

    pub fn merge_config(mut self, value: serde_json::Value) -> Self {
        self.config = merge(self.config, value, vec![]);
        self
    }

    pub fn enable_outcomes(mut self) -> Self {
        self.config = merge(self.config, outcomes_enabled_config(), vec![]);
        self
    }

    pub fn build(self) -> Relay<'a, U> {
        let config = Config::from_json_value(self.config).unwrap();
        let relay_bin = get_relay_binary().unwrap();

        let mut dir = TempDir::default();
        let dir = dir.create("relay");

        let credentials = load_credentials(&config, &dir);

        self.upstream
            .insert_known_relay(credentials.id, credentials.public_key);

        let process = BackgroundProcess::new(
            relay_bin.as_path().to_str().unwrap(),
            &["-c", dir.as_path().to_str().unwrap(), "run"],
        );

        let server_address = config.listen_addr();

        // We need this delay before we start sending to relay.
        std::thread::sleep(Duration::from_millis(500));

        Relay {
            _process: process,
            server_address,
            _config: Arc::new(config),
            _client: reqwest::Client::new(),
            upstream_dsn: self.upstream.internal_error_dsn(),
            upstream: self.upstream,
        }
    }
}

pub struct Relay<'a, U: Upstream> {
    server_address: SocketAddr,
    _process: BackgroundProcess,
    _config: Arc<Config>,
    _client: reqwest::Client,
    upstream_dsn: String,
    upstream: &'a U,
}

impl<'a, U: Upstream> Relay<'a, U> {
    pub fn server_address(&self) -> SocketAddr {
        self.server_address
    }

    pub fn get_dsn(&self, public_key: ProjectKey) -> String {
        let x = self.server_address();
        let host = x.ip();
        let port = x.port();

        format!("http://{public_key}:@{host}:{port}/42")
    }

    pub fn new(upstream: &'a U) -> Relay<'a, U> {
        Self::builder(upstream).build()
    }

    pub fn builder(upstream: &U) -> RelayBuilder<U> {
        let host = "127.0.0.1".into();
        let port = random_port();
        let url = upstream.url();
        let internal_error_dsn = upstream.internal_error_dsn();

        let config = default_opts(url, internal_error_dsn, port, host);

        RelayBuilder { config, upstream }
    }

    fn url(&self) -> String {
        format!(
            "http://{}:{}",
            self.server_address.ip(),
            self.server_address.port()
        )
    }

    pub fn get_auth_header(&self, dsn_key: ProjectKey) -> String {
        format!(
            "Sentry sentry_version=5, sentry_timestamp=1535376240291, sentry_client=rust-node/2.6.3, sentry_key={}",
            dsn_key
        )
    }

    pub fn envelope_url(&self, project_id: ProjectId) -> String {
        let endpoint = format!("/api/{}/envelope/", project_id);
        format!("{}{}", self.url(), endpoint)
    }

    pub fn send_envelope_to_url(&self, envelope: RawEnvelope, url: &str) -> Response {
        use reqwest::header::HeaderValue;

        let mut headers = HeaderMap::new();
        headers.insert(
            "Content-Type",
            HeaderValue::from_static("application/x-sentry-envelope"),
        );

        let dsn_key = self.public_dsn_key(envelope.project_id);
        headers.insert(
            "X-Sentry-Auth",
            HeaderValue::from_str(&self.get_auth_header(dsn_key)).unwrap(),
        );

        let data = envelope.serialize();

        // Add additional headers from envelope if necessary
        for (key, value) in envelope.http_headers.iter() {
            headers.insert(
                HeaderName::from_bytes(key.as_bytes()).unwrap(),
                HeaderValue::from_str(value).unwrap(),
            );
        }

        Runtime::new().unwrap().block_on(async {
            reqwest::Client::new()
                .post(url)
                .headers(headers)
                .body(data)
                .send()
                .await
                .unwrap()
        })
    }

    pub fn send_envelope(&self, envelope: RawEnvelope) {
        let url = self.envelope_url(envelope.project_id);
        self.send_envelope_to_url(envelope, &url);
    }
}

fn get_relay_binary() -> Result<PathBuf, Box<dyn std::error::Error>> {
    let version = "latest";
    if version == "latest" {
        return Ok(std::env::var("RELAY_BIN")
            .map_or_else(|_| "../target/debug/relay".into(), PathBuf::from)
            .canonicalize()
            .expect("Failed to get absolute path"));
    };

    let filename = match env::consts::OS {
        "linux" => "relay-Linux-x86_64",
        "macos" => "relay-Darwin-x86_64",
        "windows" => "relay-Windows-x86_64.exe",
        _ => panic!("Unsupported OS"),
    };

    let download_path = PathBuf::from(format!(
        "target/relay_releases_cache/{}_{}",
        filename, version
    ));

    if !Path::new(&download_path).exists() {
        let download_url = format!(
            "https://github.com/getsentry/relay/releases/download/{}/{}",
            version, filename
        );

        let client = reqwest::blocking::Client::new();
        let mut request = client.get(download_url);

        if let Ok(token) = env::var("GITHUB_TOKEN") {
            request = request.bearer_auth(token);
        }

        let response = request.send()?.error_for_status()?;

        // Adjusted part: Read the entire response body at once.
        let content = response.bytes()?;

        fs::create_dir_all(Path::new(&download_path).parent().unwrap())?;
        let mut file = File::create(&download_path)?;

        // Write the entire content to the file.
        file.write_all(&content)?;

        let mut perms = fs::metadata(&download_path)?.permissions();
        perms.set_mode(0o700); // UNIX-specific; for Windows, you'll need a different approach
        fs::set_permissions(&download_path, perms)?;
    }

    Ok(download_path)
}

fn load_credentials(config: &Config, relay_dir: &Path) -> Credentials {
    let relay_bin = get_relay_binary().unwrap();
    let config_path = relay_dir.join("config.yml");

    std::fs::write(config_path.as_path(), config.to_yaml_string().unwrap()).unwrap();

    let output = std::process::Command::new(relay_bin.as_path())
        .arg("-c")
        .arg(config_path.parent().unwrap())
        .arg("credentials")
        .arg("generate")
        .output()
        .unwrap();

    if !output.status.success() {
        dbg!(&output);
        panic!("Command execution failed");
    }
    let credentials_path = relay_dir.join("credentials.json");

    let credentials_str = std::fs::read_to_string(credentials_path).unwrap();
    serde_json::from_str(&credentials_str).expect("Failed to parse JSON")
}
